QMontgomeryUInt not serializable

I was tying to create a bloq example with dtype=QMontgomeryUInt when I recieved this error

```
bloq_ex = BloqExample(name='cmodadd_example', bloq_cls=<class 'qualtran.bloqs.mod_arithmetic.mod_addition.CModAdd'>, generalizer=<function <lambda> at 0x7f686ba19800>)

    def assert_bloq_example_serializes(bloq_ex: BloqExample) -> None:
        """Assert that the BloqExample has consistent serialization.
    
        This function asserts that the given bloq can be serialized to a proto format and the
        corresponding proto can be deserialized back to a bloq which is equal to the original
        bloq.
    
        If the given Bloq cannot be serialized / deserialized OR if the deserialized Bloq is not
        equal to the given Bloq, then the result is `FAIL`. If the roundtrip succeeds, the result
        is `PASS`.
    
        Returns:
            None if the assertions are satisfied.
    
        Raises:
            BloqCheckException if any assertions are violated.
        """
        from qualtran.serialization.bloq import bloqs_from_proto, bloqs_to_proto
    
        bloq = bloq_ex.make()
    
        try:
            bloq_lib = bloqs_to_proto(bloq)
        except Exception as e:
>           raise BloqCheckException.fail('Serialization Failed:\n' + str(e)) from e
E           qualtran.testing.BloqCheckException: Serialization Failed:
E           Data type <class 'qualtran._infra.data_types.QMontgomeryUInt'> is not recognized. It must be of one of the following subtypes: QBit, QAny, QInt, QIntOnesComp, QUInt, BoundedQUInt, QFixedPoint

qualtran/testing.py:579: BloqCheckException
```

--- a/qualtran/protos/data_types.proto
+++ b/qualtran/protos/data_types.proto
@@ -37,6 +37,10 @@ message QUInt {
     optional IntOrSympy bitsize = 1;
 }
 
+message QMontgomeryUInt {
+    optional IntOrSympy bitsize = 1;
+}
+
 message BoundedQUInt {
     optional IntOrSympy bitsize = 1;
     optional IntOrSympy iteration_length = 2;
@@ -57,5 +61,6 @@ message QDataType {
     QUInt quint = 5;
     BoundedQUInt bounded_quint = 6;
     QFxp qfxp = 7;
+    QMontgomeryUInt qmontgomery_uint = 8;
   }
 }

--- a/qualtran/protos/data_types_pb2.py
+++ b/qualtran/protos/data_types_pb2.py
@@ -15,7 +15,7 @@ _sym_db = _symbol_database.Default()
 from qualtran.protos import args_pb2 as qualtran_dot_protos_dot_args__pb2
 
 
-DESCRIPTOR = _descriptor_pool.Default().AddSerializedFile(b'\n qualtran/protos/data_types.proto\x12\x08qualtran\x1a\x1aqualtran/protos/args.proto\"\x06\n\x04QBit\">\n\x04QAny\x12*\n\x07\x62itsize\x18\x01 \x01(\x0b\x32\x14.qualtran.IntOrSympyH\x00\x88\x01\x01\x42\n\n\x08_bitsize\">\n\x04QInt\x12*\n\x07\x62itsize\x18\x01 \x01(\x0b\x32\x14.qualtran.IntOrSympyH\x00\x88\x01\x01\x42\n\n\x08_bitsize\"F\n\x0cQIntOnesComp\x12*\n\x07\x62itsize\x18\x01 \x01(\x0b\x32\x14.qualtran.IntOrSympyH\x00\x88\x01\x01\x42\n\n\x08_bitsize\"?\n\x05QUInt\x12*\n\x07\x62itsize\x18\x01 \x01(\x0b\x32\x14.qualtran.IntOrSympyH\x00\x88\x01\x01\x42\n\n\x08_bitsize\"\x90\x01\n\x0c\x42oundedQUInt\x12*\n\x07\x62itsize\x18\x01 \x01(\x0b\x32\x14.qualtran.IntOrSympyH\x00\x88\x01\x01\x12\x33\n\x10iteration_length\x18\x02 \x01(\x0b\x32\x14.qualtran.IntOrSympyH\x01\x88\x01\x01\x42\n\n\x08_bitsizeB\x13\n\x11_iteration_length\"\x98\x01\n\x04QFxp\x12*\n\x07\x62itsize\x18\x01 \x01(\x0b\x32\x14.qualtran.IntOrSympyH\x00\x88\x01\x01\x12+\n\x08num_frac\x18\x02 \x01(\x0b\x32\x14.qualtran.IntOrSympyH\x01\x88\x01\x01\x12\x13\n\x06signed\x18\x03 \x01(\x08H\x02\x88\x01\x01\x42\n\n\x08_bitsizeB\x0b\n\t_num_fracB\t\n\x07_signed\"\x97\x02\n\tQDataType\x12\x1e\n\x04qbit\x18\x01 \x01(\x0b\x32\x0e.qualtran.QBitH\x00\x12\x1e\n\x04qany\x18\x02 \x01(\x0b\x32\x0e.qualtran.QAnyH\x00\x12\x1e\n\x04qint\x18\x03 \x01(\x0b\x32\x0e.qualtran.QIntH\x00\x12\x30\n\x0eqint_ones_comp\x18\x04 \x01(\x0b\x32\x16.qualtran.QIntOnesCompH\x00\x12 \n\x05quint\x18\x05 \x01(\x0b\x32\x0f.qualtran.QUIntH\x00\x12/\n\rbounded_quint\x18\x06 \x01(\x0b\x32\x16.qualtran.BoundedQUIntH\x00\x12\x1e\n\x04qfxp\x18\x07 \x01(\x0b\x32\x0e.qualtran.QFxpH\x00\x42\x05\n\x03valb\x06proto3')
+DESCRIPTOR = _descriptor_pool.Default().AddSerializedFile(b'\n qualtran/protos/data_types.proto\x12\x08qualtran\x1a\x1aqualtran/protos/args.proto\"\x06\n\x04QBit\">\n\x04QAny\x12*\n\x07\x62itsize\x18\x01 \x01(\x0b\x32\x14.qualtran.IntOrSympyH\x00\x88\x01\x01\x42\n\n\x08_bitsize\">\n\x04QInt\x12*\n\x07\x62itsize\x18\x01 \x01(\x0b\x32\x14.qualtran.IntOrSympyH\x00\x88\x01\x01\x42\n\n\x08_bitsize\"F\n\x0cQIntOnesComp\x12*\n\x07\x62itsize\x18\x01 \x01(\x0b\x32\x14.qualtran.IntOrSympyH\x00\x88\x01\x01\x42\n\n\x08_bitsize\"?\n\x05QUInt\x12*\n\x07\x62itsize\x18\x01 \x01(\x0b\x32\x14.qualtran.IntOrSympyH\x00\x88\x01\x01\x42\n\n\x08_bitsize\"I\n\x0fQMontgomeryUInt\x12*\n\x07\x62itsize\x18\x01 \x01(\x0b\x32\x14.qualtran.IntOrSympyH\x00\x88\x01\x01\x42\n\n\x08_bitsize\"\x90\x01\n\x0c\x42oundedQUInt\x12*\n\x07\x62itsize\x18\x01 \x01(\x0b\x32\x14.qualtran.IntOrSympyH\x00\x88\x01\x01\x12\x33\n\x10iteration_length\x18\x02 \x01(\x0b\x32\x14.qualtran.IntOrSympyH\x01\x88\x01\x01\x42\n\n\x08_bitsizeB\x13\n\x11_iteration_length\"\x98\x01\n\x04QFxp\x12*\n\x07\x62itsize\x18\x01 \x01(\x0b\x32\x14.qualtran.IntOrSympyH\x00\x88\x01\x01\x12+\n\x08num_frac\x18\x02 \x01(\x0b\x32\x14.qualtran.IntOrSympyH\x01\x88\x01\x01\x12\x13\n\x06signed\x18\x03 \x01(\x08H\x02\x88\x01\x01\x42\n\n\x08_bitsizeB\x0b\n\t_num_fracB\t\n\x07_signed\"\xce\x02\n\tQDataType\x12\x1e\n\x04qbit\x18\x01 \x01(\x0b\x32\x0e.qualtran.QBitH\x00\x12\x1e\n\x04qany\x18\x02 \x01(\x0b\x32\x0e.qualtran.QAnyH\x00\x12\x1e\n\x04qint\x18\x03 \x01(\x0b\x32\x0e.qualtran.QIntH\x00\x12\x30\n\x0eqint_ones_comp\x18\x04 \x01(\x0b\x32\x16.qualtran.QIntOnesCompH\x00\x12 \n\x05quint\x18\x05 \x01(\x0b\x32\x0f.qualtran.QUIntH\x00\x12/\n\rbounded_quint\x18\x06 \x01(\x0b\x32\x16.qualtran.BoundedQUIntH\x00\x12\x1e\n\x04qfxp\x18\x07 \x01(\x0b\x32\x0e.qualtran.QFxpH\x00\x12\x35\n\x10qmontgomery_uint\x18\x08 \x01(\x0b\x32\x19.qualtran.QMontgomeryUIntH\x00\x42\x05\n\x03valb\x06proto3')
 
 _globals = globals()
 _builder.BuildMessageAndEnumDescriptors(DESCRIPTOR, _globals)
@@ -32,10 +32,12 @@ if not _descriptor._USE_C_DESCRIPTORS:
   _globals['_QINTONESCOMP']._serialized_end=280
   _globals['_QUINT']._serialized_start=282
   _globals['_QUINT']._serialized_end=345
-  _globals['_BOUNDEDQUINT']._serialized_start=348
-  _globals['_BOUNDEDQUINT']._serialized_end=492
-  _globals['_QFXP']._serialized_start=495
-  _globals['_QFXP']._serialized_end=647
-  _globals['_QDATATYPE']._serialized_start=650
-  _globals['_QDATATYPE']._serialized_end=929
+  _globals['_QMONTGOMERYUINT']._serialized_start=347
+  _globals['_QMONTGOMERYUINT']._serialized_end=420
+  _globals['_BOUNDEDQUINT']._serialized_start=423
+  _globals['_BOUNDEDQUINT']._serialized_end=567
+  _globals['_QFXP']._serialized_start=570
+  _globals['_QFXP']._serialized_end=722
+  _globals['_QDATATYPE']._serialized_start=725
+  _globals['_QDATATYPE']._serialized_end=1059
 # @@protoc_insertion_point(module_scope)

--- a/qualtran/protos/data_types_pb2.pyi
+++ b/qualtran/protos/data_types_pb2.pyi
@@ -108,6 +108,24 @@ class QUInt(google.protobuf.message.Message):
 global___QUInt = QUInt
 
 @typing.final
+class QMontgomeryUInt(google.protobuf.message.Message):
+    DESCRIPTOR: google.protobuf.descriptor.Descriptor
+
+    BITSIZE_FIELD_NUMBER: builtins.int
+    @property
+    def bitsize(self) -> qualtran.protos.args_pb2.IntOrSympy: ...
+    def __init__(
+        self,
+        *,
+        bitsize: qualtran.protos.args_pb2.IntOrSympy | None = ...,
+    ) -> None: ...
+    def HasField(self, field_name: typing.Literal["_bitsize", b"_bitsize", "bitsize", b"bitsize"]) -> builtins.bool: ...
+    def ClearField(self, field_name: typing.Literal["_bitsize", b"_bitsize", "bitsize", b"bitsize"]) -> None: ...
+    def WhichOneof(self, oneof_group: typing.Literal["_bitsize", b"_bitsize"]) -> typing.Literal["bitsize"] | None: ...
+
+global___QMontgomeryUInt = QMontgomeryUInt
+
+@typing.final
 class BoundedQUInt(google.protobuf.message.Message):
     DESCRIPTOR: google.protobuf.descriptor.Descriptor
 
@@ -173,6 +191,7 @@ class QDataType(google.protobuf.message.Message):
     QUINT_FIELD_NUMBER: builtins.int
     BOUNDED_QUINT_FIELD_NUMBER: builtins.int
     QFXP_FIELD_NUMBER: builtins.int
+    QMONTGOMERY_UINT_FIELD_NUMBER: builtins.int
     @property
     def qbit(self) -> global___QBit: ...
     @property
@@ -187,6 +206,8 @@ class QDataType(google.protobuf.message.Message):
     def bounded_quint(self) -> global___BoundedQUInt: ...
     @property
     def qfxp(self) -> global___QFxp: ...
+    @property
+    def qmontgomery_uint(self) -> global___QMontgomeryUInt: ...
     def __init__(
         self,
         *,
@@ -197,9 +218,10 @@ class QDataType(google.protobuf.message.Message):
         quint: global___QUInt | None = ...,
         bounded_quint: global___BoundedQUInt | None = ...,
         qfxp: global___QFxp | None = ...,
+        qmontgomery_uint: global___QMontgomeryUInt | None = ...,
     ) -> None: ...
-    def HasField(self, field_name: typing.Literal["bounded_quint", b"bounded_quint", "qany", b"qany", "qbit", b"qbit", "qfxp", b"qfxp", "qint", b"qint", "qint_ones_comp", b"qint_ones_comp", "quint", b"quint", "val", b"val"]) -> builtins.bool: ...
-    def ClearField(self, field_name: typing.Literal["bounded_quint", b"bounded_quint", "qany", b"qany", "qbit", b"qbit", "qfxp", b"qfxp", "qint", b"qint", "qint_ones_comp", b"qint_ones_comp", "quint", b"quint", "val", b"val"]) -> None: ...
-    def WhichOneof(self, oneof_group: typing.Literal["val", b"val"]) -> typing.Literal["qbit", "qany", "qint", "qint_ones_comp", "quint", "bounded_quint", "qfxp"] | None: ...
+    def HasField(self, field_name: typing.Literal["bounded_quint", b"bounded_quint", "qany", b"qany", "qbit", b"qbit", "qfxp", b"qfxp", "qint", b"qint", "qint_ones_comp", b"qint_ones_comp", "qmontgomery_uint", b"qmontgomery_uint", "quint", b"quint", "val", b"val"]) -> builtins.bool: ...
+    def ClearField(self, field_name: typing.Literal["bounded_quint", b"bounded_quint", "qany", b"qany", "qbit", b"qbit", "qfxp", b"qfxp", "qint", b"qint", "qint_ones_comp", b"qint_ones_comp", "qmontgomery_uint", b"qmontgomery_uint", "quint", b"quint", "val", b"val"]) -> None: ...
+    def WhichOneof(self, oneof_group: typing.Literal["val", b"val"]) -> typing.Literal["qbit", "qany", "qint", "qint_ones_comp", "quint", "bounded_quint", "qfxp", "qmontgomery_uint"] | None: ...
 
 global___QDataType = QDataType

--- a/qualtran/serialization/data_types.py
+++ b/qualtran/serialization/data_types.py
@@ -60,7 +60,7 @@ def data_type_to_proto(data: QDType) -> data_types_pb2.QDataType:
             f"Data type {type(data)} is not recognized."
             " It must be of one of the following subtypes: QBit, "
             "QAny, QInt, QIntOnesComp, QUInt, BoundedQUInt, "
-            "QFxp"
+            "QFxp, QMontgomeryUInt"
         )
 
 

--- a/qualtran/serialization/data_types.py
+++ b/qualtran/serialization/data_types.py
@@ -11,7 +11,17 @@
 #  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 #  See the License for the specific language governing permissions and
 #  limitations under the License.
-from qualtran import BoundedQUInt, QAny, QBit, QDType, QFxp, QInt, QIntOnesComp, QUInt
+from qualtran import (
+    BoundedQUInt,
+    QAny,
+    QBit,
+    QDType,
+    QFxp,
+    QInt,
+    QIntOnesComp,
+    QMontgomeryUInt,
+    QUInt,
+)
 from qualtran.protos import data_types_pb2
 from qualtran.serialization.args import int_or_sympy_from_proto, int_or_sympy_to_proto
 
@@ -29,6 +39,10 @@ def data_type_to_proto(data: QDType) -> data_types_pb2.QDataType:
         return data_types_pb2.QDataType(qint_ones_comp=data_types_pb2.QIntOnesComp(bitsize=bitsize))
     elif isinstance(data, QUInt):
         return data_types_pb2.QDataType(quint=data_types_pb2.QUInt(bitsize=bitsize))
+    elif isinstance(data, QMontgomeryUInt):
+        return data_types_pb2.QDataType(
+            qmontgomery_uint=data_types_pb2.QMontgomeryUInt(bitsize=bitsize)
+        )
     elif isinstance(data, BoundedQUInt):
         iteration_length = int_or_sympy_to_proto(data.iteration_length)
         return data_types_pb2.QDataType(
@@ -46,7 +60,7 @@ def data_type_to_proto(data: QDType) -> data_types_pb2.QDataType:
             f"Data type {type(data)} is not recognized."
             " It must be of one of the following subtypes: QBit, "
             "QAny, QInt, QIntOnesComp, QUInt, BoundedQUInt, "
-            "QFixedPoint"
+            "QFxp"
         )
 
 
@@ -66,6 +80,9 @@ def data_type_from_proto(serialized: data_types_pb2.QDataType) -> QDType:
     elif serialized.HasField('quint'):
         bitsize = int_or_sympy_from_proto(serialized.quint.bitsize)
         return QUInt(bitsize=bitsize)
+    elif serialized.HasField('qmontgomery_uint'):
+        bitsize = int_or_sympy_from_proto(serialized.qmontgomery_uint.bitsize)
+        return QMontgomeryUInt(bitsize=bitsize)
     elif serialized.HasField('bounded_quint'):
         bitsize = int_or_sympy_from_proto(serialized.bounded_quint.bitsize)
         iteration_length = int_or_sympy_from_proto(serialized.bounded_quint.iteration_length)
@@ -79,5 +96,5 @@ def data_type_from_proto(serialized: data_types_pb2.QDataType) -> QDType:
             f"Data type {type(serialized)} is not recognized."
             " It must be of one of the following subtypes: QBit, "
             "QAny, QInt, QIntOnesComp, QUInt, BoundedQUInt, "
-            "QFixedPoint"
+            "QFxp, QMontgomeryUInt"
         )

--- a/qualtran/serialization/data_types_test.py
+++ b/qualtran/serialization/data_types_test.py
@@ -17,7 +17,18 @@ from typing import Union
 import pytest
 import sympy
 
-from qualtran import BoundedQUInt, QAny, QBit, QDType, QFxp, QInt, QIntOnesComp, QUInt
+from qualtran import (
+    BoundedQUInt,
+    QAny,
+    QBit,
+    QDType,
+    QFxp,
+    QInt,
+    QIntOnesComp,
+    QMontgomeryUInt,
+    QUInt,
+)
+from qualtran._infra.data_types import QMontgomeryUInt
 from qualtran.serialization.data_types import data_type_from_proto, data_type_to_proto
 
 
@@ -37,6 +48,7 @@ def test_basic_data_types(num_qbits: Union[int, sympy.Expr]):
     round_trip_qdt(QInt(num_qbits))
     round_trip_qdt(QAny(num_qbits))
     round_trip_qdt(QIntOnesComp(num_qbits))
+    round_trip_qdt(QMontgomeryUInt(num_qbits))
     round_trip_qdt(QUInt(num_qbits))
 
 

